extensions/rich_text_components/components.py

```diff
@@ -0,0 +1,164 @@
+    """Class for Collapsible component."""
+
+    customization_args = {
+        'content-with-value': objects.Html,
+        'heading-with-value': objects.UnicodeString
+    }
+
+    @classmethod
+    def validate(cls, value_dict):
+        """Validates Collapsible component."""
+        super(Collapsible, cls).validate(value_dict)
+        content = value_dict['content-with-value']
+        inner_soup = bs4.BeautifulSoup(
+            content.encode(encoding='utf-8'),
+            'html.parser')
+        collapsible_count = len(inner_soup.findAll(
+            name='oppia-noninteractive-collapsible'))
+        tabs_count = len(inner_soup.findAll(
+            name='oppia-noninteractive-tabs'))
+        if collapsible_count or tabs_count:
+            raise Exception('Nested tabs and collapsible')
+
+
+class Image(BaseComponent):
+    """Class for Image component."""
+
+    customization_args = {
+        'filepath-with-value': objects.Filepath,
+        'alt-with-value': objects.UnicodeString,
+        'caption-with-value': objects.UnicodeString
+    }
+
+    @classmethod
+    def validate(cls, value_dict):
+        """Validates Image component."""
+        super(Image, cls).validate(value_dict)
+        filename_re = r'^[A-Za-z0-9+/]*\.((png)|(jpeg)|(gif)|(jpg))$'
+        filepath = value_dict['filepath-with-value']
+        if not re.match(filename_re, filepath):
+            raise Exception('Invalid filepath')
+
+
+class Link(BaseComponent):
+    """Class for Link component."""
+
+    customization_args = {
+        'url-with-value': objects.SanitizedUrl,
+        'text-with-value': objects.UnicodeString
+    }
+
+
+class Math(BaseComponent):
+    """Class for Math component."""
+
+    customization_args = {
+        'raw_latex-with-value': objects.MathLatexString
+    }
+
+
+class Tabs(BaseComponent):
+    """Class for Tabs component."""
+
+    customization_args = {
+        'content': objects.Html,
+        'title': objects.UnicodeString
+    }
+
+    @classmethod
+    def validate(cls, value_dict):
+        """Validates Tab component."""
+        if 'tab_contents-with-value' in value_dict:
+            tab_contents = value_dict['tab_contents-with-value']
+            if isinstance(tab_contents, list) and (
+                    all(isinstance(item, dict) for item in (
+                        tab_contents))):
+                for tab_content in tab_contents:
+                    super(Tabs, cls).validate(tab_content)
+                    inner_soup = bs4.BeautifulSoup(
+                        tab_content['content'].encode(encoding='utf-8'),
+                        'html.parser')
+                    collapsible = inner_soup.findAll(
+                        name='oppia-noninteractive-collapsible')
+                    collapsible_count = len(collapsible)
+                    tabs = inner_soup.findAll(
+                        name='oppia-noninteractive-tabs')
+                    tabs_count = len(tabs)
```

Comment:
Just inline len(collapsible) and len(tabs) below, and remove these vars. In general, if you use a var only once, it's more conceptual effort for the reader to keep it in their memory.

Reply:
Done!